velox/docs/develop/testing/fuzzer.rst

```diff
@@ -164,21 +164,27 @@ input vector and expression to files and replay these later.
 
 2. Run Expression Runner using generated files.
 
-``--repro_persist_path <path/to/directory>`` flag tells the Fuzzer to save
-input vector and expression SQL to files in the specified directory and print
-out the exact paths. Fuzzer uses :doc:`VectorSaver <../debugging/vector-saver>` for storing vectors on disk
-while preserving encodings.
-
-ExpressionRunner takes a path to input vector, path to expression SQL and
-"mode" and evaluates the specified expression on the specified data.
+``--repro_persist_path <path/to/directory>`` flag tells the Fuzzer to save the
+input vector, initial result vector, expression SQL, and other relevant data to files in a new directory saved within
+the specified directory. It also prints out the exact paths for these. Fuzzer uses :doc:`VectorSaver <../debugging/vector-saver>`
+for storing vectors on disk while preserving encodings.
+
+ExpressionRunner needs at the very least a path to input vector and path to expression SQL to run.
+However, you might need more files to reproduce the issue. All of which will be present in the directory
+that the fuzzer test generated. You can directly point the ExpressionRunner to that directory using --input_dir
+where it will pick up all the files automatically or you can specify each explicitly using other startup flags.
 ExpressionRunner supports the following flags:
 
+* ``--input_dir`` directory path where all input files (required to reproduce a failure) that are generated by the Fuzzer are expected to reside. ExpressionRunner will automatically pick up all the files from this folder unless they are explicitly specified via their respective startup flag.
+
 * ``--input_path`` path to input vector that was created by the Fuzzer
 
 * ``--sql_path`` path to expression SQL that was created by the Fuzzer
 
 * ``--complex_constant_path`` optional path to complex constants that aren't accurately expressable in SQL (Array, Map, Structs, ...). This is used with SQL file to reproduce the exact expression, not needed when the expression doesn't contain complex constants.
 
+* ``--lazy_column_list_path`` optional path for the file stored on-disk which contains a vector of column indices that specify which columns of the input row vector should be wrapped in lazy. This is used when the failing test included input columns that were lazy vector.
```

Comment:
Do we need this as a separate flag? Why can't this be included in the VectorSaver dump?

Reply:
The VectorSaver dump for a lazy vector would serialize the loaded vector. This would prevent us from playing around with the input and is useful when we don't know what the full data was in case of actual in-production failures.

However, for the expressionFuzzer the way lazy vector generation works is that:
- an input row vector is created
- some of its columns are wrapped in lazy before feeding it to common path
- the simplified path receives the non-lazy input
- If both paths fail with the same error then they are re-tried with the expression wrapped in TRY and the exact same columns are wrapped in lazy when feeding it to the common path
- This necessitates the need of preserving the input row vector and wrapping the exact same columns on-demand.

So the approach I took was to keep track of which columns we initially decide to wrap in lazy and then write it out in a separate file. On serialization, this would preserve the input vector in its entirely and still allow us the repro the crash.

Reply:
Thank you for explaining. It bothers me that adding LV support in this manner doesn't seem to uncover new bugs. At the same time the fuzzer logic around LV is pretty complicated. I feel testing LV by creating TableScan -> Project plans would be simpler in terms of code and more powerful in terms of finding bugs and testing "production" scenarios.

Reply:
Its definitely a bit disconcerting that it did not uncover bugs apart from the very early ones in dictionary loading. The LV generated by the fuzzer should behave exactly like one generated from a scan but there might be subtleties that we are missing. I'll hack together something that runs the fuzzer with a table scan.